backend/ml/inputs.py

```diff
@@ -41,63 +46,75 @@ def ratings_properties(self) -> pd.DataFrame:
             * `user_id`: int
             * `entity_id`: int or str
             * `is_public`: bool
-            * `is_trusted`: bool
-            * `is_supertrusted`: bool
+            * `is_scaling_calibration_user`: bool
+            * `trust_score`: float
         """
         raise NotImplementedError
 
 
 class MlInputFromPublicDataset(MlInput):
-    def __init__(self, csv_file):
-        self.public_dataset = pd.read_csv(csv_file)
-        self.public_dataset.rename(
-            {"video_a": "entity_a", "video_b": "entity_b"}, axis=1, inplace=True
-        )
-        self.public_dataset["user_id"], self.user_indices = self.public_dataset[
-            "public_username"
-        ].factorize()
+    def __init__(self, dataset_zip: Union[str, BinaryIO]):
+        if isinstance(dataset_zip, str) and (
+            dataset_zip.startswith("http://") or dataset_zip.startswith("https://")
+        ):
+            dataset_zip, _headers = urlretrieve(dataset_zip)  # nosec B310
+
+        with zipfile.ZipFile(dataset_zip) as zip_file:
+            zip_root_dir = next(path for path in zipfile.Path(zip_file).iterdir() if path.is_dir())
+
+            with (zip_root_dir / "comparisons.csv").open() as comparison_file:
+                self.comparisons = pd.read_csv(comparison_file)
```

Comment:
When I try it on the real dataset, I have an exception `UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 3576: character maps to <undefined>`.
Explicitly setting `encoding='utf8'` doesn't work either.

Reply:
I can't reproduce the issue. Are you using the versions as specified in "backend/requirements.txt"?

Here with Python 3.9.16, Pandas 1.4.1 :
```bash
$ python manage.py shell
Python 3.9.16 (main, Dec  7 2022, 00:00:00) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from  ml.inputs import MlInputFromPublicDataset

In [2]: ml_input = MlInputFromPublicDataset("https://api.tournesol.app/exports/all/")

In [3]: ml_input.ratings_properties
Out[3]: 
       user_id    entity_id  is_public  trust_score  is_scaling_calibration_user
0         1237  RdAoJhs1yR0       True      0.00000                        False
1         4841  hOWB7o6XUSU       True      0.00000                        False
2          611  TuWrejZE8i8       True      1.00000                         True
3         2196  Isa-qo_vXnw       True      0.00000                        False
4         2892  cOGB3WmylnY       True      1.00000                         True
...        ...          ...        ...          ...                          ...
51124     1459  4_Le9FOsO74       True      0.00000                        False
51125     3083  EFmPGvzfMZI       True      0.00000                        False
51126      150  3AXupc7oE-g       True      0.80000                         True
51127     1888  mL-6oJC2tcg       True      0.97147                         True
51128     5096  GvS2qik3Mr8       True      0.00000                        False

[51129 rows x 5 columns]

In [4]: ml_input.get_comparisons()
Out[4]: 
        user_id     entity_a     entity_b             criteria  score  weight
0          1129  tkiOzcSya18  1jFeYXgarsM  largely_recommended    -10       1
1          1129  tkiOzcSya18  1jFeYXgarsM          reliability    -10       1
2          1129  tkiOzcSya18  1jFeYXgarsM             pedagogy    -10       1
3          1129  tkiOzcSya18  1jFeYXgarsM           importance    -10       1
4          1129  tkiOzcSya18  1jFeYXgarsM      layman_friendly    -10       1
...         ...          ...          ...                  ...    ...     ...
437873     1066  7i1Bq6pr120  AFNAUEKfjpo        better_habits     -1       1
437874     1066  7i1Bq6pr120  AFNAUEKfjpo        backfire_risk     -2       1
437875     5430  Wxp__MZgwAI  d8ntTETk2F0  largely_recommended    -10       1
437876     3416  gKhOJqaD6Pw  2bI7s_d5D8Y  largely_recommended     -8       1
437877     3416  eUB6zkW-e5Y  2bI7s_d5D8Y  largely_recommended     -9       1

[437878 rows x 6 columns]
```

Reply:
I have Python 3.9.12 and pandas 1.4.1.
I also tried with Python 3.11 and reproduced the issue. I suspect that it comes from working on Windows, and from working on a Zip file.
It's similar to what's described here (https://stackoverflow.com/questions/68470744/unicodedecodeerror-utf-8-for-pandas-read-csv-from-folder-nested-zip-file), except the solutions don't work for me.

Reply:
Strange. Would it help to open the .csv files in binary mode?
```diff
diff --git a/backend/ml/inputs.py b/backend/ml/inputs.py
index 77c3160d..972abdb6 100644
--- a/backend/ml/inputs.py
+++ b/backend/ml/inputs.py
@@ -62,13 +62,13 @@ class MlInputFromPublicDataset(MlInput):
         with zipfile.ZipFile(dataset_zip) as zip_file:
             zip_root_dir = next(path for path in zipfile.Path(zip_file).iterdir() if path.is_dir())
 
-            with (zip_root_dir / "comparisons.csv").open() as comparison_file:
+            with (zip_root_dir / "comparisons.csv").open(mode="rb") as comparison_file:
                 self.comparisons = pd.read_csv(comparison_file)
                 self.comparisons.rename(
                     {"video_a": "entity_a", "video_b": "entity_b"}, axis=1, inplace=True
                 )
 
-            with (zip_root_dir / "users.csv").open() as users_file:
+            with (zip_root_dir / "users.csv").open(mode="rb") as users_file:
                 self.users = pd.read_csv(users_file)
                 self.users.index.name = "user_id"
```

Reply:
yes it works !

Reply:
Ok, fixed in 75d762c2

Regarding how to use this class to make experimentation easier, I've just opened #1370.